core/utils/syscallthread.h

```diff
@@ -0,0 +1,443 @@
+   * re-enable SIG_THREAD_EXIT after pushing a defer.  (If you
+   * are going to exit you need not bother.)
+   */
+  template <typename T = std::enable_if<!ReliableSignals, void>>
+  typename T::type PopDefer() {
+    if (--defer_count_ == 0) {
+      ThreadSetSigmask(&InternalMasks().GetExitMask());
+    }
+  }
+
+  /*
+   * Call this on the thread, from any other thread, to request
+   * termination (exit) of the running thread.  Optionally, wait
+   * for actual termination.
+   *
+   * Does nothing if the thread was never started, or is already
+   * terminated.  Note, however, that Terminate(true) will
+   * wait for the termination to complete, after an earlier
+   * Terminate(false).
+   */
+  void Terminate(enum WaitType waittype = Wait) {
+    if (state_ == TSNotStarted) {
+      // If never started, there is nothing to terminate.
+      return;
+    }
+
+    // We should signal the thread if:
+    // - it's gotten past TSReady (but this test would race
+    //   with the thread itself, so we just assume it has), and
+    // - no one else asked it to exit yet, and
+    // - it's not already on its way out, or done.
+    // (if someone else already asked, that someone-else also
+    // kicked off any signalling required).
+    bool send_signal = !exit_requested_ && state_ < TSExiting;
+    exit_requested_ = true;
+
+    if (send_signal) {
+      if (ReliableSignals) {
+        KickThread(&thread_);
+      } else {
+        assert(!knock_thread_.joinable());
+        knock_thread_ = std::thread([this]() {
+          while (this->state_ < TSExiting) {
+            KickThread(&this->thread_);
+            // Use nanosleep so that SIG_THREAD_EXIT will interrupt us.
+            // The documentation says that sleep_for waits for the full
+            // time (presumably using a loop with nanosleep calls).
+            //
+            // std::this_thread::sleep_for(std::chrono::milliseconds(250));
+            struct timespec delay = {.tv_sec = 0, .tv_nsec = 250 * 1000 * 1000};
+            nanosleep(&delay, nullptr);
+          }
+        });
+      }
+    }
+
+    if (waittype == Wait) {
+      WaitFor();
+    }
+  }
+
+  /*!
+   * If you want to verify that the job has finished, call WaitFor.
+   * Note, however, that this may result in a scheduling yield (it
+   * potentially calls join()).
+   *
+   * Note that WaitFor() DOES NOT request that the thread terminate;
+   * to do that, call Terminate() first.  If you call wait too soon
+   * you could wait forever.
+   */
+  void WaitFor() {
+    // If the thread was already joined (redundant Terminate)
+    // there's nothing more to do with that part.
+    if (thread_.joinable()) {
+      thread_.join();
+    }
+    if (knock_thread_.joinable()) {
+      // Kick the knock thread now, to make it finish early.
+      KickThread(&knock_thread_);
+      knock_thread_.join();
+    }
+  }
+
+  bool IsExitRequested() { return exit_requested_; }
```

Comment:
(minor) should be `bool IsExitRequested() const`, likewise for `Done` just below - though a const instance of `*this` makes little sense for either one anyway.